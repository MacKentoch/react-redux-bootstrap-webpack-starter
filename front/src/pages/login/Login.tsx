import React, { useState, useEffect, useCallback } from 'react';
import { RouteComponentProps } from 'react-router';
import Button from 'reactstrap/lib/Button';
import Row from 'reactstrap/lib/Row';
import Col from 'reactstrap/lib/Col';
import auth from '../../services/auth';
import FadeInEntrance from '../../components/fadeInEntrance';
import { MappedDispatchToProps, MappedStateToProps, OwnProps } from './index';

// #region types
export type Props = RouteComponentProps &
  MappedStateToProps &
  MappedDispatchToProps &
  OwnProps;
// #endregion

function Login({
  location,
  isLogging,
  isFetching,
  disconnectUser,
  logUserIfNeeded,
  history,
}: Props) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  useEffect(() => {
    // on mount disconnect (clear previous user auth and info):
    disconnectUser();
  }, []);

  // #region form inputs change callbacks
  const handlesOnEmailChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      if (event) {
        event.preventDefault();
        setEmail(event.target.value.trim());
      }
    },
    [],
  );

  const handlesOnPasswordChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      if (event) {
        event.preventDefault();
        setPassword(event.target.value.trim());
      }
    },
    [],
  );
  // #endregion

  // #region on login button click callback
  const handlesOnLogin = useCallback(
    async (event: React.MouseEvent<HTMLButtonElement>) => {
      event && event.preventDefault();

      try {
        const response = await logUserIfNeeded(email, password);
        const {
          data: { token, user },
        } = response.payload;

        auth.setToken(token);
        auth.setUserInfo(user);

        // @ts-ignore
        const { from } = location.state || { from: { pathname: '/' } };

        history.replace(from); // back to Home
      } catch (error) {
        /* eslint-disable no-console */
        console.log('login went wrong..., error: ', error);
        /* eslint-enable no-console */
      }
    },
    [history, location, email, password],
  );
  // #endregion

  // #region on go back home button click callback
  const goHome = useCallback(
    (event: React.MouseEvent<HTMLButtonElement>) => {
      event && event.preventDefault();
      history.push({ pathname: '/' });
    },
    [history],
  );
  // #endregion

  return (
    <FadeInEntrance>
      <div className="content">
        <Row>
          <Col md={{ size: 4, offset: 4 }} xs={{ size: 10, offset: 1 }}>
            <form className="form-horizontal" noValidate>
              <fieldset>
                <legend>Login</legend>
                <div className="form-group">
                  <label
                    htmlFor="inputEmail"
                    className="col-lg-2 control-label"
                  >
                    Email
                  </label>
                  <Col lg={12}>
                    <input
                      type="text"
                      className="form-control"
                      id="inputEmail"
                      placeholder="Email"
                      autoComplete="username email"
                      value={email}
                      onChange={handlesOnEmailChange}
                    />
                  </Col>
                </div>

                <div className="form-group">
                  <label
                    htmlFor="inputPassword"
                    className="col-lg-2 control-label"
                  >
                    Password
                  </label>
                  <Col lg={12}>
                    <input
                      type="password"
                      className="form-control"
                      id="inputPassword"
                      placeholder="Password"
                      autoComplete="current-password"
                      value={password}
                      onChange={handlesOnPasswordChange}
                    />
                  </Col>
                </div>

                <div className="form-group">
                  <Col lg={{ size: 12 }}>
                    <Button
                      className="login-button btn-block"
                      color="primary"
                      disabled={isLogging}
                      onClick={handlesOnLogin}
                    >
                      {isLogging ? (
                        <span>
                          login in... &nbsp;
                          <i className="fa fa-spinner fa-pulse fa-fw" />
                        </span>
                      ) : (
                        <span>Login</span>
                      )}
                    </Button>
                  </Col>
                </div>
              </fieldset>
            </form>
          </Col>
        </Row>
        <Row>
          <Col md={{ size: 4, offset: 4 }} xs={{ size: 10, offset: 1 }}>
            <div className="pull-right">
              <Button className="btn-block" onClick={goHome}>
                back to home
              </Button>
            </div>
          </Col>
        </Row>
      </div>
    </FadeInEntrance>
  );
}

Login.defaultProps = {
  isFetching: false,
  isLogging: false,
};

Login.displayName = 'Login';

export default Login;
