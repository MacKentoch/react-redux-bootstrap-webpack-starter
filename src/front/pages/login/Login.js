// @flow

// #region imports
import React, { PureComponent } from 'react';
import {
  type Match,
  type Location,
  type RouterHistory,
} from 'react-router-dom';
import Button from 'reactstrap/lib/Button';
import Row from 'reactstrap/lib/Row';
import Col from 'reactstrap/lib/Col';
import auth from '../../services/auth';
import { type UserAuthActions } from '../../types/redux/userAuth';
// #endregion

// #region flow types
export type Props = {
  // react-router 4:
  match: Match,
  location: Location,
  history: RouterHistory,

  // userAuth:
  isAuthenticated: boolean,
  isFetching: boolean,
  isLogging: boolean,
} & UserAuthActions;

export type State = {
  email: string,
  password: string,
};
// #endregion

class Login extends PureComponent<Props, State> {
  static defaultProps = {
    isFetching: false,
    isLogging: false,
  };

  state = {
    email: '',
    password: '',
  };

  // #region lifecycle methods
  componentDidMount() {
    const { disconnectUser } = this.props;

    disconnectUser(); // diconnect user: remove token and user info
  }

  render() {
    const { email, password } = this.state;

    const { isLogging } = this.props;

    return (
      <div className="content">
        <Row>
          <Col md={{ size: 4, offset: 4 }} xs={{ size: 10, offset: 1 }}>
            <form className="form-horizontal" noValidate>
              <fieldset>
                <legend>Login</legend>

                <div className="form-group">
                  <label
                    htmlFor="inputEmail"
                    className="col-lg-2 control-label"
                  >
                    Email
                  </label>
                  <Col lg={12}>
                    <input
                      type="text"
                      className="form-control"
                      id="inputEmail"
                      placeholder="Email"
                      autoComplete="username email"
                      value={email}
                      onChange={this.handlesOnEmailChange}
                    />
                  </Col>
                </div>

                <div className="form-group">
                  <label
                    htmlFor="inputPassword"
                    className="col-lg-2 control-label"
                  >
                    Password
                  </label>
                  <Col lg={12}>
                    <input
                      type="password"
                      className="form-control"
                      id="inputPassword"
                      placeholder="Password"
                      autoComplete="current-password"
                      value={password}
                      onChange={this.handlesOnPasswordChange}
                    />
                  </Col>
                </div>

                <div className="form-group">
                  <Col lg={{ size: 12 }}>
                    <Button
                      className="login-button btn-block"
                      color="primary"
                      disabled={isLogging}
                      onClick={this.handlesOnLogin}
                    >
                      {isLogging ? (
                        <span>
                          login in... &nbsp;
                          <i className="fa fa-spinner fa-pulse fa-fw" />
                        </span>
                      ) : (
                        <span>Login</span>
                      )}
                    </Button>
                  </Col>
                </div>
              </fieldset>
            </form>
          </Col>
        </Row>
        <Row>
          <Col md={{ size: 4, offset: 4 }} xs={{ size: 10, offset: 1 }}>
            <div className="pull-right">
              <Button className="btn-block" onClick={this.goHome}>
                back to home
              </Button>
            </div>
          </Col>
        </Row>
      </div>
    );
  }
  // #endregion

  // #region form inputs change callbacks
  handlesOnEmailChange = (event: SyntheticEvent<>) => {
    if (event) {
      event.preventDefault();
      // should add some validator before setState in real use cases
      this.setState({ email: event.target.value.trim() });
    }
  };

  handlesOnPasswordChange = (event: SyntheticEvent<>) => {
    if (event) {
      event.preventDefault();
      // should add some validator before setState in real use cases
      this.setState({ password: event.target.value.trim() });
    }
  };
  // #endregion

  // #region on login button click callback
  handlesOnLogin = async (event: SyntheticEvent<>) => {
    if (event) {
      event.preventDefault();
    }

    const { history, logUserIfNeeded } = this.props;

    const { email, password } = this.state;

    const userLogin = {
      login: email,
      password: password,
    };

    try {
      const response = await logUserIfNeeded(userLogin);
      const {
        data: { token, user },
      } = response.payload;

      auth.setToken(token);
      auth.setUserInfo(user);

      history.push({ pathname: '/' }); // back to Home
    } catch (error) {
      /* eslint-disable no-console */
      console.log('login went wrong..., error: ', error);
      /* eslint-enable no-console */
    }
  };
  // #endregion

  // #region on go back home button click callback
  goHome = (event: SyntheticEvent<>) => {
    if (event) {
      event.preventDefault();
    }

    const { history } = this.props;

    history.push({ pathname: '/' });
  };
  // #endregion
}

export default Login;
