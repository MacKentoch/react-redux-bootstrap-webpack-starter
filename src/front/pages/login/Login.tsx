import React, { useState, useEffect } from 'react';
import { RouteComponentProps } from 'react-router';
import Button from 'reactstrap/lib/Button';
import Row from 'reactstrap/lib/Row';
import Col from 'reactstrap/lib/Col';
import auth from '../../services/auth';
import FadeInEntrance from '../../components/fadeInEntrance';
import { MappedDispatchToProps, MappedStateToProps, OwnProps } from './index';

// #region types
export type Props = {} & RouteComponentProps &
  MappedStateToProps &
  MappedDispatchToProps &
  OwnProps;
// #endregion

function Login({
  isLogging,
  isFetching,
  disconnectUser,
  logUserIfNeeded,
  history,
}: Props) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  useEffect(() => {
    // on mount disconnect (clear previous user auth and info):
    disconnectUser();
  }, []);

  // #region form inputs change callbacks
  const handlesOnEmailChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event) {
      event.preventDefault();
      setEmail(event.target.value.trim());
    }
  };

  const handlesOnPasswordChange = (
    event: React.ChangeEvent<HTMLInputElement>,
  ) => {
    if (event) {
      event.preventDefault();
      setPassword(event.target.value.trim());
    }
  };
  // #endregion

  // #region on login button click callback
  const handlesOnLogin = async (event: React.MouseEvent<HTMLButtonElement>) => {
    event && event.preventDefault();

    try {
      const response = await logUserIfNeeded(email, password);
      console.log('"response: ', response);
      const {
        data: { token, user },
      } = response.payload;

      auth.setToken(token);
      auth.setUserInfo(user);

      history.push({ pathname: '/' }); // back to Home
    } catch (error) {
      /* eslint-disable no-console */
      console.log('login went wrong..., error: ', error);
      /* eslint-enable no-console */
    }
  };
  // #endregion

  // #region on go back home button click callback
  const goHome = (event: React.MouseEvent<HTMLButtonElement>) => {
    event && event.preventDefault();
    history.push({ pathname: '/' });
  };
  // #endregion

  return (
    <FadeInEntrance>
      <div className="content">
        <Row>
          <Col md={{ size: 4, offset: 4 }} xs={{ size: 10, offset: 1 }}>
            <form className="form-horizontal" noValidate>
              <fieldset>
                <legend>Login</legend>
                <div className="form-group">
                  <label
                    htmlFor="inputEmail"
                    className="col-lg-2 control-label"
                  >
                    Email
                  </label>
                  <Col lg={12}>
                    <input
                      type="text"
                      className="form-control"
                      id="inputEmail"
                      placeholder="Email"
                      autoComplete="username email"
                      value={email}
                      onChange={handlesOnEmailChange}
                    />
                  </Col>
                </div>

                <div className="form-group">
                  <label
                    htmlFor="inputPassword"
                    className="col-lg-2 control-label"
                  >
                    Password
                  </label>
                  <Col lg={12}>
                    <input
                      type="password"
                      className="form-control"
                      id="inputPassword"
                      placeholder="Password"
                      autoComplete="current-password"
                      value={password}
                      onChange={handlesOnPasswordChange}
                    />
                  </Col>
                </div>

                <div className="form-group">
                  <Col lg={{ size: 12 }}>
                    <Button
                      className="login-button btn-block"
                      color="primary"
                      disabled={isLogging}
                      onClick={handlesOnLogin}
                    >
                      {isLogging ? (
                        <span>
                          login in... &nbsp;
                          <i className="fa fa-spinner fa-pulse fa-fw" />
                        </span>
                      ) : (
                        <span>Login</span>
                      )}
                    </Button>
                  </Col>
                </div>
              </fieldset>
            </form>
          </Col>
        </Row>
        <Row>
          <Col md={{ size: 4, offset: 4 }} xs={{ size: 10, offset: 1 }}>
            <div className="pull-right">
              <Button className="btn-block" onClick={goHome}>
                back to home
              </Button>
            </div>
          </Col>
        </Row>
      </div>
    </FadeInEntrance>
  );
}

Login.defaultProps = {
  isFetching: false,
  isLogging: false,
};

Login.displayName = 'Login';

export default Login;
