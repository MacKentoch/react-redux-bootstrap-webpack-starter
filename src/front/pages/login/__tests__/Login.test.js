// @flow

// #region imports
import React from 'react';
import renderer from 'react-test-renderer'; // needed both for snpashot testing but also to prevent errors from enzyme
import { MemoryRouter } from 'react-router';
import Login from '../Login';
import { type Props } from '../Login';
// #endregion

describe('Login page', () => {
  it('renders as expected', () => {
    const props: Props = {
      isFetching: false,
      isLogging: false,
      isAuthenticated: false,
      logUserIfNeeded: () => {},
      disconnectUser: () => {},
    };

    const component = renderer
      .create(
        <div>
          <MemoryRouter>
            <Login {...props} />
          </MemoryRouter>
        </div>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
