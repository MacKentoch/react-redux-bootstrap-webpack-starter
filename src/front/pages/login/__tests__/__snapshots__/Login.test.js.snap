// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page renders as expected 1`] = `
<div>
  <div
    className="content"
  >
    <div
      className="row"
    >
      <div
        className="col-10 offset-1 col-md-4 offset-md-4"
      >
        <form
          className="form-horizontal"
          noValidate={true}
        >
          <fieldset>
            <legend>
              Login
            </legend>
            <div
              className="form-group"
            >
              <label
                className="col-lg-2 control-label"
                htmlFor="inputEmail"
              >
                Email
              </label>
              <div
                className="col-lg-12"
              >
                <input
                  autoComplete="username email"
                  className="form-control"
                  id="inputEmail"
                  onChange={[Function]}
                  placeholder="Email"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-lg-2 control-label"
                htmlFor="inputPassword"
              >
                Password
              </label>
              <div
                className="col-lg-12"
              >
                <input
                  autoComplete="current-password"
                  className="form-control"
                  id="inputPassword"
                  onChange={[Function]}
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
            </div>
            <div
              className="form-group"
            >
              <div
                className="col-lg-12"
              >
                <button
                  className="login-button btn-block btn btn-primary"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <span>
                    Login
                  </span>
                </button>
              </div>
            </div>
          </fieldset>
        </form>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-10 offset-1 col-md-4 offset-md-4"
      >
        <div
          className="pull-right"
        >
          <button
            className="btn-block btn btn-secondary"
            onClick={[Function]}
            type="button"
          >
            back to home
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
